### PR TITLE
feat(3464) separate matrix and batch states for settlement

### PR DESCRIPTION
**Pull request by koekiebox**
https://github.com/mojaloop/project/issues/3464

# Changes include:
- Support multiple currencies for a dynamic matrix
- Settlement model parameter is optional
- Separate the matrix status from the batch status
- Add operation to mark a matrix as `outOfSync` when a batch belonging to the matrix is updated
- Better unit tests (increased scenario coverage)
- include the statuses of the batch as lookup criteria

### CODE REVIEW

File: packages/domain-lib/test/unit/index.test.ts [hunk 1]
```diff
@@ -1288,16 +1288,229 @@ describe("Settlements BC [Domain] - Unit Tests", () => {
...
+			[batchId]
+		);
+		expect(matrixId).toBeDefined();
+
+		// close:
+		await aggregate.closeSettlementMatrix(securityContext, matrixId);
+		batch = await settleBatchRepo.getBatch(batchId);
+		expect(batch).toBeDefined();
+		expect(batch!.id).toEqual(batchId);
+		expect(batch!.state).toEqual('CLOSED');
```
Comment (tselit) [inline packages/domain-lib/test/unit/index.test.ts:1342]: Then we can show here (after ln1342) that once the batch state is `CLOSED` then no more transactions can be added to the batch.

**Review by tselit — CHANGES_REQUESTED** [review]
Hi @koekiebox 
Thank you, thank you! 
Please check out my thoughts & suggestions, and then tell me if we should huddle about any point.

File: packages/domain-lib/test/unit/index.test.ts [hunk 2]
```diff
@@ -1288,16 +1288,229 @@ describe("Settlements BC [Domain] - Unit Tests", () => {
...
+		);
+		expect(matrixId).toBeDefined();
+		await aggregate.lockSettlementMatrixForAwaitingSettlement(securityContext, matrixId);
+		await aggregate.settleSettlementMatrix(securityContext, matrixId);
+		batch = await settleBatchRepo.getBatch(batchId);
+		expect(batch).toBeDefined();
+		expect(batch!.id).toEqual(batchId);
+		expect(batch!.state).toEqual('SETTLED');
+
+		const matrixSettled = await aggregate.getSettlementMatrix(securityContext, matrixId);
```
[the remark this answers is not in the document]
Comment (koekiebox) [reply packages/domain-lib/test/unit/index.test.ts:1388]: @tselit correct. The only states for a matrix is now: 
- `IDLE` for when a matrix is awaiting input/actions
- `BUSY` when a matrix is currently being updated
- `FINALIZED` when a matrix has been settled (all batches settled)

File: packages/domain-lib/test/unit/index.test.ts [hunk 3]
```diff
@@ -1288,16 +1288,229 @@ describe("Settlements BC [Domain] - Unit Tests", () => {
...
+		expect(batch!.id).toEqual(batchId);
+		expect(batch!.state).toEqual('CLOSED');
+		// now we lock the batches via the matrix in order to settle the matrix, but we need a new matrix:
+		matrixId = await aggregate.createStaticSettlementMatrix(
+			securityContext,
+			null, // matrix-id
+			[batchId]
+		);
+		expect(matrixId).toBeDefined();
+		await aggregate.lockSettlementMatrixForAwaitingSettlement(securityContext, matrixId);
```
[the remark this answers is not in the document]
Comment (koekiebox) [reply packages/domain-lib/test/unit/index.test.ts:1381]: Resolved. ✅

File: packages/domain-lib/test/unit/index.test.ts [hunk 4]
```diff
@@ -1288,16 +1288,229 @@ describe("Settlements BC [Domain] - Unit Tests", () => {
...
-		expect(matrixNoLock!.state).toEqual('AWAITING_SETTLEMENT');
+		expect(matrixNoLock!.state).toEqual('FINALIZED');
 		expect(matrixNoLock!.type).toEqual('STATIC');
 		expect(matrixNoLock!.participantBalances.length).toEqual(2);
 		expect(matrixNoLock!.batches.length).toEqual(1);
 	});
 
-	test("test matrix state machine", async () => {
-		// TODO
+	test("test matrix state machine - positive", async () => {
```
[the remark this answers is not in the document]
Comment (koekiebox) [reply packages/domain-lib/test/unit/index.test.ts:1305]: Thank you @tselit . Additional test cases have now been added. 
The complete list of tests include:
image

New tests include:
image

File: packages/domain-lib/test/unit/index.test.ts [hunk 5]
```diff
@@ -1288,16 +1288,229 @@ describe("Settlements BC [Domain] - Unit Tests", () => {
...
+		expect(batch!.state).toEqual('OPEN');
+
+		// create a settlement matrix with batches:
+		let matrixId = await aggregate.createStaticSettlementMatrix(
+			securityContext,
+			null, // matrix-id
+			[batchId]
+		);
+		expect(matrixId).toBeDefined();
+
```
[the remark this answers is not in the document]
Comment (koekiebox) [reply packages/domain-lib/test/unit/index.test.ts:1332]: Thank you. Good idea. Added.

File: packages/domain-lib/test/unit/index.test.ts [hunk 6]
```diff
@@ -1288,16 +1288,229 @@ describe("Settlements BC [Domain] - Unit Tests", () => {
...
+			settlementModel: 'STATE_MACHINE'
+		};
+		const batchId: string = await aggregate.handleTransfer(securityContext, reqTransferDto);
+		expect(batchId).toBeDefined();
+
+		// Initial state for a batch is [OPEN]:
+		let batch = await settleBatchRepo.getBatch(batchId);
+		expect(batch).toBeDefined();
+		expect(batch!.id).toEqual(batchId);
+		expect(batch!.state).toEqual('OPEN');
```
[the remark this answers is not in the document]
Comment (koekiebox) [reply packages/domain-lib/test/unit/index.test.ts:1323]: Fixed.

File: packages/domain-lib/test/unit/index.test.ts (hunk 1 above)
Comment (koekiebox) [reply packages/domain-lib/test/unit/index.test.ts:1338]: Fixed.

**Comment by pedrosousabarreto** [comment]
Looks great!! 
Matching UI changes already in place for version 0.1.23 (building)

**Comment by koekiebox** [comment]
🦭 of approval given.

**Review by pedrosousabarreto — APPROVED** [review]
Looks good to me!

**Pull request merged**